ksp_plugin/plugin.cpp

```diff
@@ -215,89 +216,124 @@ RenderedTrajectory<World> Plugin::RenderedVesselTrajectory(
+RenderedTrajectory<World> Plugin::RenderedPrediction(
+    not_null<
+        Transforms<Barycentric, Rendering, Barycentric>*> const transforms,
+    Position<World> const& sun_world_position) {
+  CHECK(!initializing_);
+  if (!has_predictions()) {
+    return RenderedTrajectory<World>();
   }
+  Trajectory<Barycentric> const& actual_trajectory = *prediction_;
+  transforms_are_operating_on_predictions_ = true;
+  RenderedTrajectory<World> result =
+      RenderTrajectory(actual_trajectory,
+                       transforms->first_on_or_after(
+                           actual_trajectory,
+                           *actual_trajectory.fork_time()),
+                       transforms,
+                       sun_world_position);
+  transforms_are_operating_on_predictions_ = false;
+  return result;
+}
 
-  // Then build the apparent trajectory using the second transform.
-  auto apparent_trajectory = make_not_null_unique<Trajectory<Barycentric>>(
-                                 actual_trajectory.body<Body>());
-  for (auto intermediate_it = transforms->second(intermediate_trajectory);
-       !intermediate_it.at_end();
-       ++intermediate_it) {
-    apparent_trajectory->Append(intermediate_it.time(),
-                                intermediate_it.degrees_of_freedom());
-  }
+void Plugin::set_predicted_vessel(GUID const& vessel_guid) {
+  clear_predicted_vessel();
+  predicted_vessel_ = find_vessel_by_guid_or_die(vessel_guid).get();
+}
 
-  // Finally use the apparent trajectory to build the result.
-  auto initial_it = apparent_trajectory->first();
-  if (!initial_it.at_end()) {
-    for (auto final_it = initial_it;
-         ++final_it, !final_it.at_end();
-         initial_it = final_it) {
-      result.emplace_back(to_world(initial_it.degrees_of_freedom().position()),
-                          to_world(final_it.degrees_of_freedom().position()));
-    }
-  }
-  VLOG(1) << "Returning a " << result.size() << "-segment trajectory";
-  return result;
+void Plugin::clear_predicted_vessel() {
+  clear_predictions();
+  predicted_vessel_ = nullptr;
+}
+
+void Plugin::set_prediction_length(Time const& t) {
+  prediction_length_ = t;
+}
+
+void Plugin::set_prediction_step(Time const& t) {
+  prediction_step_ = t;
 }
 
 not_null<std::unique_ptr<Transforms<Barycentric, Rendering, Barycentric>>>
 Plugin::NewBodyCentredNonRotatingTransforms(
     Index const reference_body_index) const {
-  not_null<Celestial const*> reference_body =
+  // TODO(egg): this should be const, use a custom comparator in the map.
+  not_null<Celestial*> reference_body =
       FindOrDie(celestials_, reference_body_index).get();
   Transforms<Barycentric, Rendering, Barycentric>::
-      LazyTrajectory<Barycentric> const reference_body_prolongation =
+      LazyTrajectory<Barycentric> const
+  reference_body_prolongation_or_prediction =
+      [this, reference_body]() -> Trajectory<Barycentric> const& {
+        if (transforms_are_operating_on_predictions_) {
```

Comment:
We would still need this boolean, the prediction cannot be the same thing as the prolongation, and it cannot be the same thing as the history, for the same reason that the prolongation is forked from the history, rather than an extension thereof: timestep and prediction requirements may differ.

For the prolongation it's because we're using a much smaller timestep, for the prediction it's because we'll be using much longer ones (or sometimes shorter ones, if our prediction includes a burn).